### ISSUE TITLE
Not an issue, just a question: Live streams?

### ISSUE THREAD

**Issue by chsims1** [lmmx/beeb#13]
Interesting library here, especially with the new changes at the BBC.  Is it possible (and easy) to get a live stream of a programme (starting 'now'), rather than a complete episode?

**Comment by lmmx** [comment]
Hey Chris, thanks for your interest in the project! I wasn't aware of these new changes at the BBC, tell me more?

All I've seen that it's moved to mediaselector v6 and there are frequent API changes (as tracked by `get-iplayer` and `youtube-dl`) but I haven't followed the ins and outs closely before this project.

---

That's a great question and something I was thinking about, yes! Thanks for reminding me to get it on the to-do list :sunglasses: 

I've taken a look and I think the short answer is yes, at a first glance it's possible and very similar, but not exactly the same.

(If you have any particular output format in mind let me know by the way)

If I implemented live stream pulling (maybe distinguish it as "incremental pulling"), I'd get my processed end result much sooner (as I could begin processing the audio while the show was still on air), so I'd be interested in this too.

I haven't done it yet as this all originated as a proof of concept which then got refactored into a nicer object oriented style, but I didn't dare to change the actual approach of it much (yet!)

For live audio, the "start" of the programme is according to a scheduled timestamp rather than an official start (when you scrub, you see the clock "wall time" rather than a duration of time into the programme).

The only requirement for the feature you're requesting here is access to the M4S stream parts. This seems doable.

To explain how it works in case you're curious: I am calculating the 'count' of these segments (total duration of the programme divided by the segment length = segment count), and creating an MP4 by concatenating these segments (making the M4S part URLs by incrementing from 1 up to the segment count) after an initiator (a MPEG-DASH file).

- [`beeb.stream.urlsets`](https://github.com/lmmx/beeb/blob/master/src/beeb/stream/urlsets.py) has the `StreamUrlSet` class
- [`beeb.api.url_helpers`](https://github.com/lmmx/beeb/blob/master/src/beeb/api/url_helpers.py) has the methods `make_init_url` and `make_part_url` (to make the aforementioned MPEG-DASH initiator URL and incremented part URLs)

I've taken a look at what happens for live audio (you can see this yourself by clicking "inspect element" in the browser, then click the "Network" tab) and I see nothing unusual happening for live radio — the precise lengths vary, but are around 257,000 frames (± 1,000) exactly as for the on-demand segments.

The only difference is the source URLs come from JSON whose `vpid` (this means "programme ID for a particular version of an episode", [long story](https://smethur.st/posts/176135860)) is the radio station name (e.g. `bbc_radio_three`) rather than one for a particular episode.

Compare:

- https://open.live.bbc.co.uk/mediaselector/6/select/version/2.0/mediaset/pc/vpid/bbc_radio_three/format/json/
- https://open.live.bbc.co.uk/mediaselector/6/select/version/2.0/mediaset/pc/vpid/m000tvkz/format/json/

Obviously, this only simplifies things. As you probably noticed, I already have code in place for handling schedules, and can add a little more to target the "on air" annotated programme from the schedule (rather than say, try to calculate this against the system time or something).

Anyway, getting bogged down in the details of it: so before I look at the JSON calls involved, I'd expect that either:

- A) you can get the start segment from the programme JSON
- B) you can't, and you just have to do some calculation based on the scheduled start time(?)

So the JSON we want to get our live URLs from comes from [https://open.live.bbc.co.uk/mediaselector/6/select/version/2.0/mediaset/pc/vpid/bbc_radio_three/format/json/][1] (in the web app this URL has a `jsfunc/JS_callbacks0` on the end, but we don't need that).

[1]: https://open.live.bbc.co.uk/mediaselector/6/select/version/2.0/mediaset/pc/vpid/bbc_radio_three/format/json/

When we open this JSON, it's a list of available sources for the stream. Some of the options use the `http` protocol, and some use `https`. I chose to only use `https` streams, and I chose among the multiple options by the `priority` value:

https://github.com/lmmx/beeb/blob/26e5dc3c4ff9590d3518c7015423bbc795606f6d/src/beeb/api/json_helpers.py#L96-L113

(Woops, hit Ctrl+Enter and it posted there, forgive me I'm still writing my answer! I'll continue it as an edit)

In plain English, this means "pick the option with the smallest integer value for `priority` which uses https protocol and has `transferFormat` of "dash"."

- I think it's perhaps undocumented but significant that this will pull the lower quality stream (bitrate of 48 rather than 128), as I'm using this to downsample and then process (additionally it's faster). Maybe I should make that an explicit option...

The `href` of this JSON item gives the URL of a "[Media Presentation Description](https://en.wikipedia.org/wiki/Dynamic_Adaptive_Streaming_over_HTTP#Overview)" XML file, which is parsed with the [`MpdXml`](https://github.com/lmmx/beeb/blob/master/src/beeb/api/xml_helpers.py) class in `beeb.api.xml_helpers`.

Here's what's in the current file for BBC 3 on air right now

```xml
<?xml version="1.0" encoding="utf-8"?>
<MPD xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="urn:mpeg:dash:schema:mpd:2011" xmlns:dvb="urn:dvb:dash:dash-extensions:2014-1"
  xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 http://standards.iso.org/ittf/PubliclyAvailableStandards/MPEG-DASH_schema_files/DASH-MPD.xsd"
  type="dynamic" availabilityStartTime="1969-12-31T23:59:44Z"
  minimumUpdatePeriod="PT6H" timeShiftBufferDepth="PT6H" maxSegmentDuration="PT7S" minBufferTime="PT3.200S"
  profiles="urn:dvb:dash:profile:dvb-dash:2014,urn:dvb:dash:profile:dvb-dash:isoff-ext-live:2014"
  publishTime="2021-03-09T16:39:08">

    <UTCTiming schemeIdUri="urn:mpeg:dash:utc:http-iso:2014" value="https://time.akamai.com/?iso" />

    <BaseURL dvb:priority="1" dvb:weight="1" serviceLocation="cfs">https://as-dash-uk.live.cf.md.bbci.co.uk/pool_904/live/uk/bbc_radio_three/bbc_radio_three.isml/dash/</BaseURL>

    <Period id="1" start="PT0S">
    <AdaptationSet group="1" contentType="audio" lang="en" minBandwidth="48000" maxBandwidth="96000"
                   segmentAlignment="true" audioSamplingRate="48000" mimeType="audio/mp4" codecs="mp4a.40.5" startWithSAP="1">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <Role schemeIdUri="urn:mpeg:dash:role:2011" value="main"/>
        <SegmentTemplate timescale="48000" initialization="bbc_radio_three-$RepresentationID$.dash"
                       media="bbc_radio_three-$RepresentationID$-$Number$.m4s" startNumber="1" duration="307200"/>
        <Representation id="audio=48000" bandwidth="48000"/>
        <Representation id="audio=96000" bandwidth="96000"/>
    </AdaptationSet>

    <AdaptationSet group="1" contentType="audio" lang="en" minBandwidth="128000" maxBandwidth="320000"
                   segmentAlignment="true" audioSamplingRate="48000" mimeType="audio/mp4" codecs="mp4a.40.2" startWithSAP="1">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
        <Role schemeIdUri="urn:mpeg:dash:role:2011" value="main"/>
        <SegmentTemplate timescale="48000" initialization="bbc_radio_three-$RepresentationID$.dash"
                       media="bbc_radio_three-$RepresentationID$-$Number$.m4s" startNumber="1" duration="307200"/>
        <Representation id="audio=128000" bandwidth="128000"/>
        <Representation id="audio=320000" bandwidth="320000"/>
    </AdaptationSet>
  </Period>
</MPD>
```

The important part of this is the `BaseURL` tag:

```xml
    <BaseURL dvb:priority="1" dvb:weight="1" serviceLocation="cfs">https://as-dash-uk.live.cf.md.bbci.co.uk/pool_904/live/uk/bbc_radio_three/bbc_radio_three.isml/dash/</BaseURL>
```

Its text content is the prefix for each and every M4S stream segment. Unlike the filenames of programmes that have finished, the filenames of programmes still on air are much longer (we cannot simply increment up to the maximum segment count).

An example of some filenames on air right now (at 23:58) are:

- https://as-dash-uk.live.cf.md.bbci.co.uk/pool_904/live/uk/bbc_radio_three/bbc_radio_three.isml/dash/bbc_radio_three-audio=320000-252773429.m4s
- https://as-dash-uk.live.cf.md.bbci.co.uk/pool_904/live/uk/bbc_radio_three/bbc_radio_three.isml/dash/bbc_radio_three-audio=320000-252773430.m4s
- https://as-dash-uk.live.cf.md.bbci.co.uk/pool_904/live/uk/bbc_radio_three/bbc_radio_three.isml/dash/bbc_radio_three-audio=320000-252773431.m4s
- ... and so on

I can't figure out where those numbers come from off the top of my head, can you? E.g. `252773429` (the first part is the `$RepresentationId` which doesn't change, I can't tell when that counter resets so unsure what to count it back relative to...)

**Update**! After looking at it again, I think I figured out what the number is, which maybe gets us toward being able to create the URL:

- Given that each segment is roughly 257,000 frames and the audio sampling rate is 48kHz, then the length of each segment is roughly (257/48) = 5.35 seconds.
- Since the count is increasing with each new segment, it's clearly marking a duration from some starting point, which we can estimate
- (5.35 seconds * 252,779,700) = 1,352,371,395 seconds = 42.85 years
  - The count extends to roughly 43 years
- 2021 - 43 = 1978

Unfortunately we're about a decade off the value in the XML of `availabilityStartTime="1969-12-31T23:59:44Z"`, so I still can't tell how to calculate this...

We're in early April, so about 3.25 months are over out of 12 (so it's around 2021.28 in decimal), and that date is the very end of 1969 so we can call it 1970.0, making their difference 2021.28 - 1970,
so around 51.28 years = 1,618,273,728 seconds. If these were split evenly into 252,779,700 pieces (a count that just passed by this morning) they'd each be around 6.4 seconds long.

This seems not too far off: `maxSegmentDuration="PT7S" minBufferTime="PT3.200S"` puts a hard limit of 7 seconds on the segments (and some sort of minimum of 3.2s).

I can't tell you how I'd accurately calculate the segment numberings back to the start of a show, given the current segment count... I'll have to look more closely at what the API itself does.
